core/primitives/src/utils.rs

```diff
@@ -16,10 +16,51 @@ use crate::version::{
     ProtocolVersion, CORRECT_RANDOM_VALUE_PROTOCOL_VERSION, CREATE_HASH_PROTOCOL_VERSION,
 };
 use std::mem::size_of;
+use std::ops::Deref;
 
 /// Number of nano seconds in a second.
 const NS_IN_SECOND: u64 = 1_000_000_000;
 
+/// A data structure for tagging data as already being validated to prevent redundant work.
+pub enum MaybeValidated<T> {
+    Validated(T),
+    NotValidated(T),
+}
+
+impl<T> MaybeValidated<T> {
+    pub fn validate_with<E, F: FnOnce(&T) -> Result<bool, E>>(&self, f: F) -> Result<bool, E> {
+        match &self {
+            Self::Validated(_) => Ok(true),
+            Self::NotValidated(t) => f(t),
+        }
+    }
+
+    pub fn map<U, F: FnOnce(T) -> U>(self, f: F) -> MaybeValidated<U> {
+        match self {
+            Self::Validated(t) => MaybeValidated::Validated(f(t)),
+            Self::NotValidated(t) => MaybeValidated::NotValidated(f(t)),
+        }
+    }
+
+    pub fn unwrap(self) -> T {
```

Comment:
Can we name it differently? When reading the code it is very easy to think that `unwrap` implies that it may panic. Maybe `unbox`?

Reply:
I normally think of something that is boxed as being heap-allocated, but that is not the case here. How about `extract` instead?

Reply:
Sure that works

Reply:
Done in https://github.com/near/nearcore/pull/3689/commits/311d7b3c21355396270f496fdb5f8f43e8f49d41